packages/search/src/search.tsx

```diff
@@ -211,19 +197,24 @@ const InnerSearch = (
 	}, [] );
 
 	useUpdateEffect( () => {
+		if ( searchMode === 'on-enter' ) {
+			return;
+		}
+
 		if ( keyword ) {
-			if ( searchMode === 'when-typing' ) {
-				doSearch( keyword );
-				onSearchChange?.( keyword );
-			}
+			doSearch( keyword );
 		} else {
-			if ( searchMode === 'on-enter' ) {
-				return;
+			// explicitly bypass debouncing when keyword is empty
+			if ( delaySearch ) {
+				// Cancel any pending debounce
+				doSearch.cancel?.();
 			}
 
-			bypassOnSearchDebounce( keyword );
-			onSearchChange?.( keyword );
+			// call the callback directly without debouncing
+			onSearch?.( keyword );
 		}
+
+		onSearchChange?.( keyword );
```

Comment:
This is easier to understand when reviewing [against trunk](https://github.com/Automattic/wp-calypso/compare/b16d56427470b3f640be784fbfa620de5f7ef185...178f9b2130b78fe1321322dd242e263beeb7b0dd#diff-714aa6bf4174d2a19dc85f63e66b1cce1d33738c6b84f907fc66f8f7962dfb32R200-R203).